File: research/experiments/sweep/runs/whenwords-full-cpp-sonnet45/src/duration.cpp
#include "duration.h"
#include <stdexcept>
#include <vector>
#include <cmath>

struct Unit {
    int64_t seconds;
    const char* singularVerbose;
    const char* pluralVerbose;
    const char* compact;
};

std::string duration(int64_t seconds, const DurationOptions& options) {
    if (seconds < 0) {
        throw std::invalid_argument("Seconds must not be negative");
    }

    static const std::vector<Unit> units = {
        {31536000, "year", "years", "y"},
        {2592000, "month", "months", "mo"},
        {86400, "day", "days", "d"},
        {3600, "hour", "hours", "h"},
        {60, "minute", "minutes", "m"},
        {1, "second", "seconds", "s"}
    };

    if (seconds == 0) {
        return options.compact ? "0s" : "0 seconds";
    }

    std::vector<std::string> parts;
    int64_t remaining = seconds;

    for (size_t i = 0; i < units.size() && parts.size() < static_cast<size_t>(options.maxUnits); ++i) {
        const auto& unit = units[i];

        if (remaining >= unit.seconds) {
            int64_t count;

            // If this is the last slot, round instead of floor
            if (parts.size() + 1 >= static_cast<size_t>(options.maxUnits)) {
                count = std::round(static_cast<double>(remaining) / unit.seconds);
                remaining = 0; // Consumed all remaining
            } else {
                count = remaining / unit.seconds;
                remaining = remaining % unit.seconds;
            }

            // Format the part
            std::string part;
            if (options.compact) {
                part = std::to_string(count) + unit.compact;
            } else {
                part = std::to_string(count) + " ";
                part += (count == 1) ? unit.singularVerbose : unit.pluralVerbose;
            }

            parts.push_back(part);
        }
    }

    // Join parts
    std::string result;
    const char* separator = options.compact ? " " : ", ";

    for (size_t i = 0; i < parts.size(); ++i) {
        if (i > 0) {
            result += separator;
        }
        result += parts[i];
    }

    return result;
}

File: research/experiments/sweep/runs/whenwords-full-kotlin-sonnet45/src/main/kotlin/whenwords/Duration.kt
package whenwords

import kotlin.math.round

data class DurationOptions(
    val compact: Boolean = false,
    val maxUnits: Int = 2
)

/**
 * Formats a number of seconds as a human-readable duration string.
 *
 * @param seconds Non-negative integer (throws on negative)
 * @param options Optional formatting options (compact mode, max units)
 * @return A formatted duration string
 * @throws IllegalArgumentException if seconds is negative
 */
fun duration(seconds: Long, options: DurationOptions = DurationOptions()): String {
    if (seconds < 0) {
        throw IllegalArgumentException("Seconds must not be negative")
    }

    if (seconds == 0L) {
        return if (options.compact) "0s" else "0 seconds"
    }

    val units = listOf(
        Unit(31536000, "year", "years", "y"),
        Unit(2592000, "month", "months", "mo"),
        Unit(86400, "day", "days", "d"),
        Unit(3600, "hour", "hours", "h"),
        Unit(60, "minute", "minutes", "m"),
        Unit(1, "second", "seconds", "s")
    )

    val parts = mutableListOf<String>()
    var remaining = seconds

    for (unit in units) {
        if (parts.size >= options.maxUnits) {
            break
        }

        if (remaining >= unit.value) {
            val isLastSlot = parts.size + 1 >= options.maxUnits

            val count = if (isLastSlot) {
                // Round on the last unit (using half-up rounding)
                val division = remaining.toDouble() / unit.value
                val result = if (division - division.toLong() >= 0.5) {
                    division.toLong() + 1
                } else {
                    division.toLong()
                }
                result
            } else {
                // Floor and keep remainder
                remaining / unit.value
            }

            val label = if (options.compact) {
                unit.compact
            } else {
                if (count == 1L) unit.singular else unit.plural
            }

            parts.add(if (options.compact) "$count$label" else "$count $label")

            if (!isLastSlot) {
                remaining %= unit.value
            }
        }
    }

    return if (options.compact) {
        parts.joinToString(" ")
    } else {
        parts.joinToString(", ")
    }
}

private data class Unit(
    val value: Long,
    val singular: String,
    val plural: String,
    val compact: String
)
